android/app/src/main/java/ru/spbstu/amd/learnbraille/serial/UsbSerial.kt

```diff
@@ -0,0 +1,117 @@
+package ru.spbstu.amd.learnbraille.serial
+
+import android.app.PendingIntent
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.hardware.usb.UsbDevice
+import android.hardware.usb.UsbDeviceConnection
+import android.hardware.usb.UsbManager
+import com.felhr.usbserial.UsbSerialDevice
+import com.felhr.usbserial.UsbSerialInterface
+import ru.spbstu.amd.learnbraille.database.BrailleDots
+import timber.log.Timber
+
+class UsbSerial// TODO make this class a singleton and initialize on startup; allow multiple connection attempts
+    (usbManager: UsbManager, context: Context) {
+    private var mUsbManager: UsbManager = usbManager
+    private var invokingContext: Context = context
+    var mDevice: UsbDevice? = null
+    var mSerial: UsbSerialDevice? = null
+    var mConnection: UsbDeviceConnection? = null
+    private var birthTime: Long = System.currentTimeMillis()
+
+    companion object {
+        const val arduinoNanoVendorId = 6790
+        const val ACTION_USB_PERMISSION = "permission"
+        const val USB_SERVICE = "usb" // TODO why error: USB_SERVICE = Context.USB_SERVICE
+    }
+
+    init {
+        startUsbConnection()
+    }
+
+    fun trySend(brailleDots: BrailleDots) {
+        val data = brailleDots.toString().replace("F", "1").replace("E", "0")
+        if (System.currentTimeMillis() - birthTime > 5000)
+            sendData(data)
+    }
+
+
+    private fun startUsbConnection() {
+        val usbDevices: HashMap<String, UsbDevice>? = mUsbManager.deviceList
+        if (!usbDevices?.isEmpty()!!) {
+            var keep = true
+            usbDevices.forEach { entry ->
+                mDevice = entry.value
+                val deviceVendorId: Int? = mDevice?.vendorId
+                if (deviceVendorId == arduinoNanoVendorId) {
```

Comment:
Diagonal programming

Reply:
explain please, this tells me nothing

Reply:
Function as about action, not only about sharing code